src/D2L.CodeStyle.Analyzers/Async/BlockingAnalyzer.cs

```diff
@@ -0,0 +1,359 @@
+			ctx.RegisterCompilationEndAction(
+				ctx => WarnForUnusedBlockingAttributes(
+					ctx,
+					attr,
+					havePossiblyUnusedBlockingAttribute: havePossiblyUnusedBlockingAttribute.Keys,
+					callsSomethingBlocking: callsSomethingBlocking.Keys
+				)
+			);
+		}
+
+		private static void InspectMethodDeclaration(
+			SymbolAnalysisContext ctx,
+			INamedTypeSymbol blockingAttr,
+			Action<IMethodSymbol> noteMethodWithPossiblyUnusedBlockingAttribute
+		) {
+			var methodSymbol = (IMethodSymbol)ctx.Symbol;
+
+			// For non-async methods, look at the things we implement and
+			// whether they are blocking or not.
+			var implementedThings = methodSymbol.GetImplementedMethods();
+
+			var implementedBlockingThings = new List<IMethodSymbol>();
+			var implementedNonBlockingThings = new List<IMethodSymbol>();
+
+			foreach( var implementedMethod in implementedThings ) {
+				if( HasAttribute( implementedMethod, blockingAttr ) ) {
+					implementedBlockingThings.Add( implementedMethod );
+				} else {
+					implementedNonBlockingThings.Add( implementedMethod );
+				}
+			}
+
+			var attrData = GetAttribute( methodSymbol, blockingAttr );
+
+			if( attrData == null ) {
+				if( implementedBlockingThings.Count == 0 ) {
+					// We don't have [Blocking] and none of the things we implement are
+					// blocking, so there is nothing to do here.
+					return;
+				} else {
+					// We implement something blocking but don't have [Blocking]. That's
+					// fine: direct callers get to know that we're not actually blocking,
+					// and indirect callers have to assume the worse.
+					// Still output a diagnostic so that we can suggest a code fix in the
+					// IDE to add [Blocking].
+
+					// TODO: code fix to add [Blocking]
+
+					var decl = methodSymbol.DeclaringSyntaxReferences.First().GetSyntax() as MethodDeclarationSyntax;
+
+					ctx.ReportDiagnostic(
+						Diagnostic.Create(
+							Diagnostics.NonBlockingImplementationOfBlockingThing,
+							decl.Identifier.GetLocation(),
+							$"{methodSymbol.ContainingType.Name}.{methodSymbol.Name}",
+							$"{implementedBlockingThings[0].ContainingType.Name}.{implementedBlockingThings[0].Name}"
+						)
+					);
+					return;
+				}
+			}
+
+			// So we do have a [Blocking] attribute.
+
+			if( implementedBlockingThings.Count == 0 ) {
+				// We're not forced to have [Blocking] due to what we implement, so check
+				// if we need it at all.
+				noteMethodWithPossiblyUnusedBlockingAttribute( methodSymbol );
+			}
+
+			if( methodSymbol.IsAsync ) {
+				// TODO: code fix to remove [Blocking]
+
+				ctx.ReportDiagnostic(
+					Diagnostic.Create(
+						Diagnostics.AsyncMethodCannotBeBlocking,
+						attrData.ApplicationSyntaxReference.GetSyntax().GetLocation(),
+						methodSymbol.Name
+					)
+				);
+				return;
+			}
+
+			// Don't let us use [Blocking] if we're implementing something non-blocking,
+			// callers need to know that they may block.
+			if( implementedNonBlockingThings.Count > 0 ) {

```

Comment:
I don't think there's explicit TODO comments about it anymore, so checking:

We'll have to be able to configure, say `IDisposable.Dipose`, as `[Blocking]` and not just be relying on `HasAttribute`, right?

Reply:
Ah, found the TODO I was thinking of in the method call one